src/commands/test.rs

```diff
@@ -0,0 +1,325 @@
+        {
+            panic!("dimensions of input matrix are inconsistent")
+        }
+        if initial_len.is_none() {
+            initial_len = Some(individual_input_with_possible_values.len());
+        }
+        for (inner_index, possibility) in enumerate(individual_input_with_possible_values) {
+            if outer_index == 0 {
+                result.push(vec![possibility]);
+            } else {
+                result[inner_index].push(possibility);
+            }
+        }
+    }
+    result
+}
+
+/// Compute an iterator of [`Run`]s from a user provided matrix of inputs.
+///
+/// Steps to complete transformation:
+/// 1. distribute (by duplication) inner input keys to each "possible value" for
+///    that key.
+///     - this converts `IndexMap<String, Vec<Value>>` to `Vec<(String, Value)>`
+/// 2. zip groups of possible inputs with [`zip_inputs`]
+///     - this reshapes a `Vec<Vec<(String, Value)>>` (the type signature
+///       remains the same, but the nested structure changes)
+/// 3. take the cartesian product of all zipped groups
+///     - this yields items of type `Vec<Vec<(String, Value)>>`
+/// 4. flatten each item yielded
+///     - results in a `Vec<(String, value)>`
+/// 5. convert each item into a [`Run`] (which is a type alias for
+///    `IndexMap<String, Value>`)
+fn compute_runs_from_matrix(
+    matrix: impl Iterator<Item = InputMapping>,
+) -> impl Iterator<Item = Run> {
+    matrix
+        .map(|input_mapping| {
+            let mut inputs_to_zip = vec![];
+            for (key, vals) in input_mapping {
+                let mut possible_inputs = Vec::new();
+                // step 1
+                for possible_val in vals {
+                    possible_inputs.push((key.clone(), possible_val));
+                }
+                inputs_to_zip.push(possible_inputs);
+            }
+
+            let mut run_subsets = Vec::new();
+            // step 2
+            for run_subset in zip_inputs(inputs_to_zip) {
+                run_subsets.push(run_subset);
+            }
+            run_subsets
+        })
+        .multi_cartesian_product() // step 3
+        .map(|product| product.into_iter().flatten().collect()) // step 4 and 5
+}
+
+/// Performs the `test` command.
+pub async fn test(args: Args) -> CommandResult<()> {
+    let source = args.source.unwrap_or_default();
+    let source = match &source {
+        Source::Remote(_) => {
+            return Err(anyhow!("the `test` subcommand does not accept remote sources").into());
+        }
+        Source::Directory(_) | Source::File(_) => source,
+    };
+    let results = Analysis::default()
+        .add_source(source.clone())
+        .run()
+        .await
+        .map_err(CommandError::from)?;
+
+    for result in results.filter(&[&source]) {
+        let document = result.document();
+        let wdl_path = PathBuf::from(document.path().as_ref());
+        let mut yaml_path = wdl_path.with_extension("yaml");
```

Comment:
A discussion for the future, but as a developer I generally don't like to see test files cluttering up my source directory.

I would suggest that we look at a well-known location relative to the source file, like `<parent-of-source>/test/<source>.[yaml|yml]`.

Related to that, I could see paths in the test file treated as relative to the test file; for example, `fixtures/foo.bam` would be at `<parent-of-source>/test/fixtures/foo.bam`. Then one doesn't need a `$FIXTURES` variable.